### PR TITLE
add basic openbis connection and user space check

### PR DESCRIPTION
**Pull request by wow-such-code**

### CODE REVIEW

File: src/main/groovy/life/qbic/portal/sampletracking/DependencyManager.groovy [hunk 1]
```diff
@@ -116,21 +116,38 @@ class DependencyManager {
...
+        OpenBisClient openbis
+        Set<String> spaces = new HashSet<>()
+        try {
+            log.info("Trying to connect to openBIS")
+            openbis = new OpenBisClient(configManager.getDataSourceUser(), configManager.getDataSourcePassword(), configManager.getDataSourceUrl())
+            openbis.login()
+            log.info("Fetching user spaces for " + userID)
+            spaces.addAll(openbis.getUserSpaces(userID))
+        } catch (Exception e) {
+            Styles.notification("openBIS connection error",
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/portal/sampletracking/DependencyManager.groovy]: You could also just `log.error` the message and throw the Exception. The application then defaults to the default `Portlet currently not available` message which should suffice I think.

File: src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy [hunk 2]
```diff
@@ -11,32 +14,55 @@ import life.qbic.portal.sampletracking.trackinginformation.query.locations.Query
...
-        }
+      Location sampleLocation = sampleTrackingDataSource.currentSampleLocation(sampleId)
+      Sample sample = new Sample()
+      sample.setCurrentLocation(sampleLocation)
+      sample.setCode(sampleId)
+      this.querySampleOutput.publishSample(sample)
+    } catch (SampleTrackingQueryException e) {
+      log.error("Query ${sampleId} failed.", e)
+      // We pass the original error message with a generic notification to the output port
+      this.querySampleOutput.invokeOnError "Could not retrieve information for sample $sampleId. ${e.message}"
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy:49]: Do you really want the user to see the Exception message?

Comment (wow-such-code) [reply src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy:49]: good point, but that was already in the code before :)

File: src/main/groovy/life/qbic/portal/sampletracking/DependencyManager.groovy (hunk 1 above)
Comment (wow-such-code) [reply src/main/groovy/life/qbic/portal/sampletracking/DependencyManager.groovy]: thanks, missed this section!

File: src/main/groovy/life/qbic/portal/sampletracking/datasources/OpenbisDataSource.groovy [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+  public void checkUserAuthorization(String sampleId) {
+    List<Sample> res = openbis.searchSampleByCode(sampleId)
+    if(res.size() == 0) {
+      throw new SampleNotInOpenbisException("User tried searching for $sampleId, but sample does not exist in openBIS. Unable to verify authorization.")
+    }
+
+    if(!userSpaces.contains(res.get(0).getSpace().getCode())) {
+      throw new OpenbisAuthorizationException("User tried searching for $sampleId, but user is not allowed to view this project.")
+    }
+  }
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/portal/sampletracking/datasources/OpenbisDataSource.groovy:34]: I am not sure if this communication using Exceptions is the way to go in the data source. 
```suggestion
  @Override
  public boolean isUserAuthorizedForSample(String sampleId) {
    List<Sample> res = openbis.searchSampleByCode(sampleId)
    if(res.size() < 1 ) {
      throw new SampleNotInOpenbisException("User tried searching for $sampleId, but sample does not exist in openBIS. Unable to verify authorization.")
    }

    return !(userSpaces.contains(res.get(0).getSpace().getCode()))
  }
```

File: src/main/groovy/life/qbic/portal/sampletracking/datasources/SampleManagementDataSource.groovy [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package life.qbic.portal.sampletracking.datasources
+
+interface SampleManagementDataSource {
+
+    void checkUserAuthorization(String sampleId)
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/portal/sampletracking/datasources/SampleManagementDataSource.groovy:5]:
```suggestion
    boolean isUserAuthorizedForSample(String sampleId)
```

File: src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy [hunk 5]
```diff
@@ -11,32 +15,43 @@ import life.qbic.portal.sampletracking.trackinginformation.query.locations.Query
...
+  QuerySample(SampleTrackingQueryDataSource dataSource, SampleManagementDataSource sampleManagementDataSource, QuerySampleOutput querySampleOutput) {
+    this.sampleTrackingDataSource = dataSource
+    this.querySampleOutput = querySampleOutput
+    this.sampleManagementDataSource = sampleManagementDataSource;
+  }
+
+  @Override
+  def querySampleById(String sampleId) {
+    try {
+      sampleManagementDataSource.checkUserAuthorization(sampleId)
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy:39]:
```suggestion
     if (sampleManagementDataSource. isUserAuthorizedForSample(sampleId) {

```

File: src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy [hunk 6]
```diff
@@ -11,32 +15,43 @@ import life.qbic.portal.sampletracking.trackinginformation.query.locations.Query
...
-        } catch (SampleTrackingQueryException e) {
-            log.error("Query ${sampleId} failed.", e)
-            // We pass the original error message with a generic notification to the output port
-            this.querySampleOutput.invokeOnError "Could not retrieve information for sample $sampleId. ${e.message}"
-        }
+      Location sampleLocation = sampleTrackingDataSource.currentSampleLocation(sampleId)
+      Sample sample = new Sample()
+      sample.setCurrentLocation(sampleLocation)
+      sample.setCode(sampleId)
+      this.querySampleOutput.publishSample(sample)
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy:45]:
```suggestion
        Location sampleLocation = sampleTrackingDataSource.currentSampleLocation(sampleId)
        Sample sample = new Sample()
        sample.setCurrentLocation(sampleLocation)
        sample.setCode(sampleId)
        this.querySampleOutput.publishSample(sample)
      } else {
      // here you can throw an Exception for example or leave it out
      }
```

File: src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy [hunk 7]
```diff
@@ -36,7 +36,8 @@ class QuerySample implements QuerySampleInput {
   @Override
   def querySampleById(String sampleId) {
     try {
-      sampleManagementDataSource.checkUserAuthorization(sampleId)
+     if (sampleManagementDataSource. isUserAuthorizedForSample(sampleId) {
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy:39]:
```suggestion
     if (sampleManagementDataSource.isUserAuthorizedForSample(sampleId) {
```

File: src/main/groovy/life/qbic/portal/sampletracking/DependencyManager.groovy [hunk 8]
```diff
@@ -116,21 +131,20 @@ class DependencyManager {
...
         } catch (Exception e) {
             log.error("Could not setup ${QueryAvailableLocations.getSimpleName()} use case", e)
         }
-
+        
         try {
-            SampleTrackingQueryDataSource trackingInfoCenter = SampleTracker.createSampleTrackingInformation(trackingServices.get(0), this.serviceUser)
-            this.querySampleInteractor = new QuerySample(trackingInfoCenter, sampleListPresenter)
+            SampleTrackingQueryDataSource trackingInfoCenter = SampleTracker.createSampleTrackingInformation(trackingServices.get(0), serviceUser)
+            this.querySampleInteractor = new QuerySample(trackingInfoCenter, sampleListPresenter, openbis, spaces)
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/portal/sampletracking/DependencyManager.groovy:142]: Neither `openbis` nor `spaces` are defined at this point. This was moved to the OpenbisDataSource, correct?

File: src/main/groovy/life/qbic/portal/sampletracking/DependencyManager.groovy [hunk 9]
```diff
@@ -58,6 +59,19 @@ class DependencyManager {
             log.error("Could not setup sample tracking service.", e)
         }
 
+        // set up openBIS connection and data management system object
+        def userID = "not logged in"
+        try {
+            userID = PortalUtils.getScreenName()
+        } catch (NullPointerException e) {
+            log.error("User not logged into Liferay. They won't be able to see samples.")
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/portal/sampletracking/DependencyManager.groovy:67]:
```suggestion
            log.error("User not logged into Liferay. They won't be able to see samples.", e)
```

File: src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy [hunk 10]
```diff
@@ -11,32 +15,45 @@ import life.qbic.portal.sampletracking.trackinginformation.query.locations.Query
 @Log4j2
 class QuerySample implements QuerySampleInput {
 
-    final SampleTrackingQueryDataSource sampleTrackingDataSource
+  final OpenBisClient openbisClient
+  final SampleTrackingQueryDataSource sampleTrackingDataSource
+  final SampleManagementDataSource sampleManagementDataSource
+  final Set<String> openbisSpaces
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy:21]:
```suggestion
```
not used

File: src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy [hunk 11]
```diff
@@ -11,32 +15,45 @@ import life.qbic.portal.sampletracking.trackinginformation.query.locations.Query
 @Log4j2
 class QuerySample implements QuerySampleInput {
 
-    final SampleTrackingQueryDataSource sampleTrackingDataSource
+  final OpenBisClient openbisClient
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy:18]:
```suggestion
```
not used

File: src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy [hunk 12]
```diff
@@ -11,32 +15,45 @@ import life.qbic.portal.sampletracking.trackinginformation.query.locations.Query
...
+  }
 
-    QuerySample(SampleTrackingQueryDataSource dataSource, QuerySampleOutput querySampleOutput) {
-        this.sampleTrackingDataSource = dataSource
-        this.querySampleOutput = querySampleOutput
-    }
+  QuerySample(SampleTrackingQueryDataSource dataSource, SampleManagementDataSource sampleManagementDataSource, QuerySampleOutput querySampleOutput) {
+    this.sampleTrackingDataSource = dataSource
+    this.querySampleOutput = querySampleOutput
+    this.sampleManagementDataSource = sampleManagementDataSource;
```
Comment (KochTobi) [inline src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy:33]:
```suggestion
    this.sampleManagementDataSource = sampleManagementDataSource
```
We don't need that :)

File: src/main/groovy/life/qbic/portal/sampletracking/datasources/OpenbisDataSource.groovy [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+
+  @Override
+  public boolean isUserAuthorizedForSample(String sampleId) {
+    List<Sample> res = openbis.searchSampleByCode(sampleId)
+    if(res.size() < 1 ) {
+      throw new SampleNotInOpenbisException("User tried searching for $sampleId, but sample does not exist in openBIS. Unable to verify authorization.")
+    }
+
+    return !userSpaces.contains(res.get(0).getSpace().getCode())
```
Comment (wow-such-code) [inline src/main/groovy/life/qbic/portal/sampletracking/datasources/OpenbisDataSource.groovy:31]: invert boolean

File: src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy [hunk 14]
```diff
@@ -11,32 +14,43 @@ import life.qbic.portal.sampletracking.trackinginformation.query.locations.Query
...
+  def querySampleById(String sampleId) {
+    try {
+      if (sampleManagementDataSource.isUserAuthorizedForSample(sampleId)) {
+        Location sampleLocation = sampleTrackingDataSource.currentSampleLocation(sampleId)
+        Sample sample = new Sample()
+        sample.setCurrentLocation(sampleLocation)
+        sample.setCode(sampleId)
+        this.querySampleOutput.publishSample(sample)
+      } else {
+        throw new OpenbisAuthorizationException("User tried searching for $sampleId, but user is not allowed to view this project.")
```
Comment (wow-such-code) [inline src/main/groovy/life/qbic/portal/sampletracking/trackinginformation/query/sample/QuerySample.groovy:43]: don't throw, just log?

File: src/main/groovy/life/qbic/portal/sampletracking/datasources/OpenbisDataSource.groovy [hunk 15]
```diff
@@ -0,0 +1,33 @@
...
+import life.qbic.portal.sampletracking.trackinginformation.query.SampleNotInOpenbisException
+import life.qbic.portal.utils.ConfigurationManager
+
+@Log4j2
+class OpenbisDataSource implements SampleManagementDataSource {
+
+  private OpenBisClient openbis;
+  private Set<String> userSpaces = new HashSet<>()
+
+  public OpenbisDataSource(ConfigurationManager configManager, String userID) {
```
Comment (wow-such-code) [inline src/main/groovy/life/qbic/portal/sampletracking/datasources/OpenbisDataSource.groovy:16]: don't use config manager, use builder pattern

**Pull request merged**